In [2]:
import re
import pandas as pd
sys.path.append('../')
import src.limpieza as lmp

In [3]:
pd.options.display.max_columns = None

In [4]:
revius = open("output1.html", "r", encoding='utf-8').readlines()

In [5]:
corte = '\ue9ce\ue9ce\ue9ce\ue9ce\ue9ce\n'

In [6]:
civitatis = [' '.join(str(i) for i in revius)]

In [7]:
print(len(civitatis))

1


In [8]:
civitatis_ = civitatis[0].split(corte)

In [9]:
civitatis_revius = [i for i in civitatis_[2::2]]

In [10]:
civitatis_revius_limpio = [i.split('\n') for i in civitatis_revius]

In [11]:
longitud = [len(i) for i in civitatis_revius_limpio]
long_ = set(longitud)
print(long_)

{4, 5, 6, 7, 8, 9, 10, 11, 12, 13}


In [12]:
for i in range(0,14):
    for a in civitatis_revius_limpio:
        if len(a)==i:
            print(a)
            break

[' 11 / Mar / 2022', ' A', ' Anónimo', ' ']
[' 10 / Mar / 2022', ' D', ' Diana Cruz Torres', ' Toa Alta, Puerto Rico', ' ']
[' 11 / Mar / 2022', ' A', ' Ainhoa', ' España', ' Viajó en pareja', ' ']
[' 11 / Mar / 2022', ' A', ' Ana', ' Madrid, España', ' Victor es encantador, te explica todo al detalle y te resuelve todas las preguntas, volvería a repetir, me a encantado.', ' Viajó con hijos mayores', ' ']
[' 11 / Mar / 2022', ' G', ' Graciela Arguelles', ' Guanajuato, Montenegro', ' Excelente recorrido ️️️️️', ' Opinión traducida. Mostrar en idioma original', ' Viajó con amigos', ' ']
[' 08 / Mar / 2022', ' A', ' Anónimo', ' 16195 opiniones', ' Mostrando 37-48', ' primera anterior', ' Página 2 Página 3 Página 4 Página 5 Página 6', ' siguiente última', ' ']
[' 10 / Feb / 2022', ' A', ' Anónimo', ' Viajó solo', ' 16195 opiniones', ' Mostrando 469-480', ' primera anterior', ' Página 38 Página 39 Página 40 Página 41 Página 42', ' siguiente última', ' ']
[' 11 / Mar / 2022', ' M', ' María J

In [13]:
civitatis_limpio_oi = lmp.pulir(civitatis_revius_limpio)
civitatis_reluciente = lmp.quitar_respuestas(civitatis_limpio_oi)


In [14]:
longitud = [len(i) for i in civitatis_reluciente]
long_ = set(longitud)
print(long_)

{2, 3, 4, 5}


In [15]:
len(civitatis_reluciente)

16231

In [16]:
dict_revius = lmp.dict_civitatis_completo(civitatis_reluciente)

In [17]:
len(dict_revius['fecha']) == len(dict_revius['nombre']) == len(dict_revius['procedencia']) == len(dict_revius['opinion']) == len(dict_revius['viajo_con'])

True

In [18]:
df_civitatis = pd.DataFrame(dict_revius)

In [19]:
df_civitatis.head(5)

fecha                     nombre      procedencia  \
0   11 / Mar / 2022                        Ana   Madrid, España   
1   11 / Mar / 2022                 Montserrat           España   
2   11 / Mar / 2022                     Ainhoa           España   
3   11 / Mar / 2022   Patricia Suárez Carrillo   Madrid, España   
4   11 / Mar / 2022                    Anónimo                    

                                             opinion                 viajo_con  
0   Victor es encantador, te explica todo al deta...   Viajó con hijos mayores  
1   Casi 3 horas de tour pero Víctor, el guía que...                Viajó solo  
2                                                              Viajó en pareja  
3   A mi pareja y a mi nos ha encantado! Nuestro ...           Viajó en pareja  
4

In [20]:
df_civitatis.shape

(16231, 5)

In [21]:
df_civitatis = lmp.values_sin_espacios(df_civitatis)

In [22]:
df_civitatis.iloc[0,0]

'11 / Mar / 2022'

In [23]:
df_civitatis['fecha'] = df_civitatis['fecha'].str.replace(' / ', '-')

In [24]:
df_civitatis['fecha'] = df_civitatis['fecha'].str.replace('Ene', '01')
df_civitatis['fecha'] = df_civitatis['fecha'].str.replace('Feb', '02')
df_civitatis['fecha'] = df_civitatis['fecha'].str.replace('Mar', '03')
df_civitatis['fecha'] = df_civitatis['fecha'].str.replace('Abr', '04')
df_civitatis['fecha'] = df_civitatis['fecha'].str.replace('May', '05')
df_civitatis['fecha'] = df_civitatis['fecha'].str.replace('Jun', '06')
df_civitatis['fecha'] = df_civitatis['fecha'].str.replace('Jul', '07')
df_civitatis['fecha'] = df_civitatis['fecha'].str.replace('Ago', '08')
df_civitatis['fecha'] = df_civitatis['fecha'].str.replace('Sep', '09')
df_civitatis['fecha'] = df_civitatis['fecha'].str.replace('Oct', '10')
df_civitatis['fecha'] = df_civitatis['fecha'].str.replace('Nov', '11')
df_civitatis['fecha'] = df_civitatis['fecha'].str.replace('Dic', '12')

In [25]:
df_civitatis.duplicated().sum()

4690

We can se that we have quite a lot of duplicates, actually around 29% of the rows seem to be replicated but, is this completely truth?

In [26]:
df_civitatis[df_civitatis.duplicated() == True]

fecha                    nombre     procedencia  \
12     11-03-2022                       Ana  Madrid, España   
13     11-03-2022                Montserrat          España   
14     11-03-2022                    Ainhoa          España   
15     11-03-2022  Patricia Suárez Carrillo  Madrid, España   
16     11-03-2022                   Anónimo                   
...           ...                       ...             ...   
16177  04-05-2019                   Anónimo          España   
16190  06-03-2019                   Anónimo          España   
16196  15-02-2019                   Anónimo          España   
16198  08-02-2019                   Anónimo          España   
16221  14-10-2018                   Anónimo          España   

                                                 opinion  \
12     Victor es encantador, te explica todo al detal...   
13     Casi 3 horas de tour pero Víctor, el guía que ...   
14                                                         
15     A mi pareja y a mi nos ha encantado! Nuestro g...   
16                                                         
...                                                  ...   
16177                                                      
16190                                                      
16196                                                      
16198                                                      
16221                                                      

                     viajo_con  
12     Viajó con hijos mayores  
13                  Viajó solo  
14             Viajó en pareja  
15             Viajó en pareja  
16                              
...                        ...  
16177                           
16190                           
16196                           
16198                           
16221                           

[4690 rows x 5 columns]

In [27]:
df_civitatis.iloc[13,3]

'Casi 3 horas de tour pero Víctor, el guía que nos ha acompañado ha sido un encanto. Hemos andado muchísimo pero ha merecido la pena porque se nota que le gusta mucho su trabajo y se ha hecho muy ameno y emocionante. Sin duda un tour 100% recomendado con este guía. Gracias por hacernos pasar una mañana increíble.'

In [28]:
df_civitatis[df_civitatis['opinion'] == 'Casi 3 horas de tour pero Víctor, el guía que nos ha acompañado ha sido un encanto. Hemos andado muchísimo pero ha merecido la pena porque se nota que le gusta mucho su trabajo y se ha hecho muy ameno y emocionante. Sin duda un tour 100% recomendado con este guía. Gracias por hacernos pasar una mañana increíble.']


fecha      nombre procedencia  \
1   11-03-2022  Montserrat      España   
13  11-03-2022  Montserrat      España   
37  11-03-2022  Montserrat      España   

                                              opinion   viajo_con  
1   Casi 3 horas de tour pero Víctor, el guía que ...  Viajó solo  
13  Casi 3 horas de tour pero Víctor, el guía que ...  Viajó solo  
37  Casi 3 horas de tour pero Víctor, el guía que ...  Viajó solo

We can be absolutely sure that the we have doubled values only if we check the column 'opinion', as it's the only one unrepeatable. On the other hand, we can easely have several people from Spain doing a tour the same day who wouldn't leave their names, so all of them will appear as 'Anonimo' and 'España' with the other columns in blank, and pandas will take them as duplicates while they actually were different clients.
So let's analize this in two different data frames. First let's take the revius with the 'opinion' column filled.

In [29]:
dict_revius_5 = lmp.dict_civitatis_5(civitatis_reluciente)

In [32]:
df_civitatis_5 = pd.DataFrame(dict_revius_5)

In [38]:
df_civitatis_5.shape

(8953, 5)

In [36]:
df_civitatis_5.head(5)

fecha                     nombre        procedencia  \
0   11 / Mar / 2022                        Ana     Madrid, España   
1   11 / Mar / 2022                 Montserrat             España   
2   11 / Mar / 2022   Patricia Suárez Carrillo     Madrid, España   
3   11 / Mar / 2022                      Saray   Alicante, España   
4   11 / Mar / 2022       Tania Carrera Garcia             España   

                                             opinion                 viajo_con  
0   Victor es encantador, te explica todo al deta...   Viajó con hijos mayores  
1   Casi 3 horas de tour pero Víctor, el guía que...                Viajó solo  
2   A mi pareja y a mi nos ha encantado! Nuestro ...           Viajó en pareja  
3                   El guia del tour, Ponce, Genial.          Viajó con amigos  
4   Mónica ha sido una guía genial, nos ha explic...          Viajó con amigos

In [37]:
df_civitatis_5.duplicated().sum()

2114

In [39]:
df_civitatis_5[df_civitatis_5.duplicated() == True]

fecha                     nombre               procedencia  \
9      11 / Mar / 2022                        Ana            Madrid, España   
10     11 / Mar / 2022                 Montserrat                    España   
11     11 / Mar / 2022   Patricia Suárez Carrillo            Madrid, España   
12     11 / Mar / 2022                      Saray          Alicante, España   
13     11 / Mar / 2022       Tania Carrera Garcia                    España   
...                ...                        ...                       ...   
8836   15 / Jun / 2017                      Pablo   Buenos Aires, Argentina   
8837   15 / Jun / 2017            Sofia Hernandez      Guatemala, Guatemala   
8838   15 / Jun / 2017                       Luis            Madrid, España   
8839   11 / Jun / 2017                      Maria            Madrid, España   
8866   25 / Sep / 2021              Rita Cantante           Porto, Portugal   

                                                opinion  \
9      Victor es encantador, te explica todo al deta...   
10     Casi 3 horas de tour pero Víctor, el guía que...   
11     A mi pareja y a mi nos ha encantado! Nuestro ...   
12                     El guia del tour, Ponce, Genial.   
13     Mónica ha sido una guía genial, nos ha explic...   
...                                                 ...   
8836    Excelente paseo y pablo nuestro guia fantastico   
8837   Me gusto mucho, amable y dispuesto a contesta...   
8838   Una guía agradable de unas dos horas en las q...   
8839   Nos ha encantado, muy amena y Pablo el guía s...   
8866   Muy interactivo, muy divertido. ¡Alex sabía m...   

                     viajo_con  
9      Viajó con hijos mayores  
10                  Viajó solo  
11             Viajó en pareja  
12            Viajó con amigos  
13            Viajó con amigos  
...                        ...  
8836           Viajó en pareja  
8837           Viajó en pareja  
8838   Viajó con hijos mayores  
8839           Viajó en pareja  
8866           Viajó en pareja  

[2114 rows x 5 columns]

In [40]:
df_civitatis_5.iloc[0, 3]

' Victor es encantador, te explica todo al detalle y te resuelve todas las preguntas, volvería a repetir, me a encantado.'

In [41]:
df_civitatis_5[df_civitatis_5['opinion'] == ' Victor es encantador, te explica todo al detalle y te resuelve todas las preguntas, volvería a repetir, me a encantado.']

fecha nombre      procedencia  \
0    11 / Mar / 2022    Ana   Madrid, España   
9    11 / Mar / 2022    Ana   Madrid, España   
26   11 / Mar / 2022    Ana   Madrid, España   

                                              opinion  \
0    Victor es encantador, te explica todo al deta...   
9    Victor es encantador, te explica todo al deta...   
26   Victor es encantador, te explica todo al deta...   

                   viajo_con  
0    Viajó con hijos mayores  
9    Viajó con hijos mayores  
26   Viajó con hijos mayores

And here we can see that that some reviews are repeated up to three times... 
Anyway, from the total number of rows, 8953, this means that 23,6% of them are duplicated, a value significantly lower than the 29% we got for whole data frame.
Now let's drop the duplicates in this one and then we will work with the reviews that don't have the 'opinion' column filled.

In [42]:
df_civitatis_5 = df_civitatis_5.drop_duplicates()
df_civitatis_5.duplicated().sum()

0

In [43]:
df_civitatis_5[df_civitatis_5['opinion'] == ' Victor es encantador, te explica todo al detalle y te resuelve todas las preguntas, volvería a repetir, me a encantado.']

fecha nombre      procedencia  \
0   11 / Mar / 2022    Ana   Madrid, España   

                                             opinion                 viajo_con  
0   Victor es encantador, te explica todo al deta...   Viajó con hijos mayores

In [31]:
df_civitatis_.duplicated().sum()

NameError: name 'df_civitatis_' is not defined

In [ ]:
df_civitatis_.shape

(11541, 5)

In [ ]:
df_civitatis['fecha'] = pd.to_datetime(df_civitatis['fecha'], format='%d-%m-%Y')

In [ ]:
df_civitatis['fecha'][0].year

2022